### PR TITLE
(feat) O3-2163: Add support for editing end visit data/time on current visits

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR provisions the adding of end date to current visits

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/53287480/ab35250d-5b7c-4a17-b400-298cefcf7d87

## Related Issue

[O3-2163](https://openmrs.atlassian.net/browse/O3-2163)

## Other

[O3-2163]: https://openmrs.atlassian.net/browse/O3-2163?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by brandones — CHANGES_REQUESTED** [review]
Thanks for this, @usamaidrsk . Can you confirm that the start visit validator on lines 151-154 make sense? I don't really understand why if there is an existing visit then it's fine if the start date is in the future.

```jsx
        (value) => {
          const today = dayjs();
          const startDate = dayjs(value);
          return displayVisitStopDateTimeFields ? true : startDate.isSameOrBefore(today, 'day');
        },
```

Please add unit tests for the end date behavior. The logic of how we decide to show these fields or not is a little funny, so it would be very good to make the intention explicit in tests.

**Comment by denniskigen** [comment]
@vasharma05 @usamaidrsk could you guys sync up to figure out how to move this forward?

**Comment by brandones** [comment]
Ping @usamaidrsk @vasharma05

**Comment by brandones** [comment]
Ping @usamaidrsk

**Review by brandones — CHANGES_REQUESTED** [review]
Looks like we have a failing E2E test, and it is being caused by this work. Please fix the E2E test.

File: packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx [hunk 1]
```diff
@@ -167,13 +169,17 @@ const StartVisitForm: React.FC<StartVisitFormProps> = ({
...
-              .refine((value) => value.match(time12HourFormatRegex), t('invalidTimeFormat', 'Invalid time format'))
-          : z.string().optional(),
-        visitStopTimeFormat: displayVisitStopDateTimeFields ? z.enum(['PM', 'AM']) : z.enum(['PM', 'AM']).optional(),
+        visitStopDate: displayVisitStopDateTimeFields && hadPreviousStopDateTime ? z.date() : z.date().optional(),
+        visitStopTime:
+          displayVisitStopDateTimeFields && hadPreviousStopDateTime
+            ? z
+                .string()
+                .refine((value) => value.match(time12HourFormatRegex), t('invalidTimeFormat', 'Invalid time format'))
+            : z.string().optional(),
```
Comment (brandones) [inline packages/esm-patient-chart-app/src/visit/visit-form/visit-form.component.tsx:178]: Should this optional one also have the `refine` clause?

**Review by brandones — APPROVED** [review]

**Pull request merged**